### PR TITLE
Feat: WebGL hit testing (Picking) alongside DOM events

### PR DESCRIPTION
**Pull request by FarhadG**

### CODE REVIEW

**Comment by FarhadG** [comment]
@michaelobriena 

Examples were tested for DOM and GL together, with and without retina displays, along with scenes comprising of large # of meshes.

File: webgl-renderers/Program.js [hunk 1]
```diff
@@ -389,13 +391,10 @@ Program.prototype.setUniforms = function (uniformNames, uniformValue) {
...
 
-        location = this.uniformLocations[name];
+        location = this.uniformLocations[name] || gl.getUniformLocation(this.program, name);
+        if (!location) continue;
 
-        if (location === null) continue;
-        if (location === undefined) {
-            location = gl.getUniformLocation(this.program, name);
-            this.uniformLocations[name] = location;
-        }
```
Comment (redwoodfavorite) [inline webgl-renderers/Program.js]: This is a rebase error.

File: webgl-renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -134,9 +137,113 @@ function WebGLRenderer(canvas, compositor) {
...
+ * Determine the alpha channel, given an x and y coordinate for the WebGL canvas.
+ *
+ * @method
+ *
+ * @param {Number} x X coordinate
+ * @param {Number} y Y coordinate
+ *
+ * @return {undefined} undefined
+ */
+WebGLRenderer.prototype.check = function check(x, y) {
```
Comment (redwoodfavorite) [inline webgl-renderers/WebGLRenderer.js]: more descriptive function name

File: webgl-renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -134,9 +137,113 @@ function WebGLRenderer(canvas, compositor) {
...
+    var gl = this.gl;
+
+    gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);
+    this.program.setUniforms(['u_clicked'], [1.0]);
```
Comment (redwoodfavorite) [inline webgl-renderers/WebGLRenderer.js]: suggest maybe 'u_pickingMode'

File: webgl-renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -184,6 +291,54 @@ WebGLRenderer.prototype.createLight = function createLight(path) {
...
+ * @param {Number} meshId Mesh ID number
+ * @param {Number} base Base for conversion
+ *
+ * @returns {Array} Encoded value
+ */
+WebGLRenderer.prototype.encodeMeshIdColor = function encodeMeshIdColor(meshId, base) {
+    var result = [];
+
+    while (meshId) {
+        var normalizedRemainder = (meshId%base) / 255;
```
Comment (redwoodfavorite) [inline webgl-renderers/WebGLRenderer.js]: var declaration above loop

**Comment by FarhadG** [comment]
Thanks for the review, @redwoodfavorite :+1: 

Suggestions were implemented and pushed.

**Comment by michaelobriena** [comment]
@FarhadG can you push up some containers showing this working for WebGL and also with both WebGL and DOM?

**Comment by FarhadG** [comment]
@michaelobriena 

Here's a container showing DOM & WebGL working together: <a href="https://github.com/FarhadG/famous-webgl-events" target="_blank">DEMO</a>

**Comment by FarhadG** [comment]
Going to put a cleaner PR in.

**Pull request closed without merging**